### ISSUE TITLE
misleading function name QgsFeature.setAttribute()/.setAttributes()

### ISSUE THREAD

**Issue by tschmetzer** [qgis/QGIS#41389]
Dear developers,

using pygis I've passed one whole day trying to figure out why I couldn't do something like 
```
# initially set values for attributes
debug_feature.setAttributes(debug_feature_attribs)
debug_layer.dataProvider().addFeature(debug_feature)
debug_layer.updateFeature(debug_feature)
debug_layer.updateExtents() 

#set a field without value
debug_feature.setAttribute(debug_feature.fieldNameIndex(feld_type), layer_type)
debug_layer.updateFeature(debug_feature)
debug_layer.updateExtents() 
```

From a usability point of view I cannot understand why I can set attributes only once in the beginning and why I shouldn't be able to call that method a second time to update it's value. I must admit I sticked only to the pygis programming docs and only in the end found the hint in the cookbook: 
https://docs.qgis.org/3.16/en/docs/pyqgis_developer_cookbook/vector.html#modify-features
Why is it necessary to make it so complicated?

I would suggest to improve the QgsFeature.setAttribute() method in order to make it work so it is **re**settable. I'm sure I am not the only one stumbling over this pitfall.
If this takes too much time and nobody wants to touch this delicate topic I suggest to update the docs and the method's name, e.g. QgsFeature.setAttribute() -> QgsFeature.setInitialAttribute()

**Comment by uclaros** [comment]
Have you started editing on the layer? Because your snippet should only work when the layer is in editing mode.
If not, you need to enclose your code in a `debug_layer.startEditing()` - `debug_layer.commitChanges()`
If you don't want to commit, you might want to use the undo buffer too by using `debug_layer.beginEditCommand('I changed an attribute')` - `debug_layer.endEditCommand()`

Your adding the feature worked without starting editing because you write directly to the underlying provider and not to the vector layer (`debug_layer.dataProvider().addFeature(debug_feature)`)

**Comment by elpaso** [comment]
not a bug

**Comment by tschmetzer** [comment]
> Have you started editing on the layer? Because your snippet should only work when the layer is in editing mode.
> If not, you need to enclose your code in a `debug_layer.startEditing()` - `debug_layer.commitChanges()`
> If you don't want to commit, you might want to use the undo buffer too by using `debug_layer.beginEditCommand('I changed an attribute')` - `debug_layer.endEditCommand()`
> 
> Your adding the feature worked without starting editing because you write directly to the underlying provider and not to the vector layer (`debug_layer.dataProvider().addFeature(debug_feature)`)

Sorry I provided only the most relevant code snippet. Yes, the layer was opened in edit mode.

**Comment by tschmetzer** [comment]
> not a bug

Tickets shouldn't be closed unilaterally without having agreed on solution. 

Why not? From a user perspective I consider it a bug. The method does not work as indicated in the docs. How can we handle this issue?

**Comment by nyalldawson** [comment]
@tschmetzer it definitely works -- it's used everywhere in qgis, and in thousands of plugins. 

We'd need a self contained example which can be run in the python console showing the failure in order to do anything here. Otherwise the only reasonable conclusion is that it's something in your particular script which is written incorrectly...

**Comment by tschmetzer** [comment]
Thanks for the valuable information that it should work!I'll come up with a complete code example showing what I'm doing wrong or where the docs might be unclear ;)Viele Grüße, Tobias Am 06.02.21, 19:38 schrieb Nyall Dawson <notifications@github.com>:

   
  @tschmetzer it definitely works -- it's used everywhere in qgis, and in thousands of plugins. 
  We'd need a self contained example which can be run in the python console showing the failure in order to do anything here. Otherwise the only reasonable conclusion is that it's something in your particular script which is written incorrectly...

**Comment by tschmetzer** [comment]
Sorry for the hassle. I must admit I'm still confused. 
- According to the pyqgis_developer_cookbook QgsFeature offers the method setAttribute() can be used with a string or an index as first argument to describe the attribute concerned. Quoting line 5 and 6 of the example at https://docs.qgis.org/3.16/en/docs/pyqgis_developer_cookbook/vector.html#add-features:
```
    feat.setAttribute('name', 'hello')
    feat.setAttribute(0, 'hello')
```

- According to pygis docs at https://qgis.org/pyqgis/master/core/QgsFeature.html#qgis.core.QgsFeature.setAttribute only an index can be used
-> Will the docs need an update?

- In the example attached I could replicate the Key Error I had gotten when I tried to access the attribute via an index. No idea why it doesn't work.
-> Any suggestions?

**Comment by nyalldawson** [comment]
> According to pygis docs at https://qgis.org/pyqgis/master/core/QgsFeature.html#qgis.core.QgsFeature.setAttribute only an index can be used
-> Will the docs need an update?

It's a styling issue on the pyqgis docs (See https://github.com/qgis/pyqgis/issues/18 ) -- the highlghted text below shows that you can use string attribute names, it's just missing the usual header format:

![image](https://user-images.githubusercontent.com/1829991/107162560-d5343200-69ef-11eb-810f-d2860a319ac8.png)

**Comment by nyalldawson** [comment]
> feature.setAttributes([[0, 'examplename1'],[1,'10'],[2,'130']])

this is wrong -- you are setting your attributes to lists of two items each! I.e. the first attribute is being set to a list value of "[0, "examplename1"]

```
print(feature.attributes())
[[0, 'examplename1'], [1, '10'], [2, '130']]
```

This should be just:
```
feature.setAttributes(['examplename1','10','130'])
```

But the actual cause of your problem is that your field handling is all over the place! After you've created the two layers with the same fields, you then add an extra field to debug_layer.  NOTE : this is added to debug_layer, NOT the data provider for debug_layer!!

So now:

[f.name() for f in feature_layer.fields()] gives  ['name', 'age', 'size']
[f.name() for f in debug_layer.fields()] gives  ['name', 'age', 'size', 'Featuretyp']
[f.name() for f in debug_layer.dataProvider().fields()] gives  ['name', 'age', 'size']

After
```
feature = QgsFeature(feature_layer.fields())
```
you have:
[f.name() for f in feature.fields()]: ['name', 'age', 'size']

After:
```
debug_feature=QgsFeature()
debug_feature.setFields(debug_layer.fields())
``
You have:
[f.name() for f in debug_feature.fields()]  : ['name', 'age', 'size', 'Featuretyp']

After:
```
debug_feature_attribs=feature.attributes()
```
debug_feature_attribs is ['examplename1', '10', '130']

You then call
debug_feature.setAttributes(debug_feature_attribs)

^^ This is the cause of your error. You're setting the attributes for debug_feature to a list of only 3 elements, whereas is should have 4 elements to match the size of the fields.

So then when you call:
```
debug_feature.setAttribute(debug_feature.fieldNameIndex(feld_type), layer_type)
```
You're calling:
debug_feature.setAttribute(3, layer_type)
Which is out of bounds for the 3 element list of attributes you just set, hence the error

What you should be doing is changing:

debug_layer.dataProvider().addFeature(debug_feature)
to
```assert debug_layer.dataProvider().addFeature(debug_feature)```
You'll see that the provider is rejecting the feature here, because it doesn't have the correct length of attributes set for the data provider. The message log window also contains a message describing why the addFeature called failed in detail.

So basically, you need to rework your field handling and insert more debugging statements into your code in order to track exactly what's occurring where!

**Comment by tschmetzer** [comment]
> But the actual cause of your problem is that your field handling is all over the place! After you've created the two layers with the same fields, you then add an extra field to debug_layer. NOTE : this is added to debug_layer, NOT the data provider for debug_layer!!
> 
> So now:
> 
> [f.name() for f in feature_layer.fields()] gives ['name', 'age', 'size']
> [f.name() for f in debug_layer.fields()] gives ['name', 'age', 'size', 'Featuretyp']
> [f.name() for f in debug_layer.dataProvider().fields()] gives ['name', 'age', 'size']

Thanks for pointing this out. The meaning of the distinction between QgsVectorLayer and QgsVectorDataProvider is now clearer to me. I understand that the commit is the sync between the two. I know there is the description in the commitChanges() method "Attempts to commit to the underlying data provider any buffered changes" but I was missing that bit of information in the class descriptions of QgsVectorLayer and QgsVectorDataProvider.

What about supplementing a short sentence like "Changes to the VectorLayer are synqued (what's the correct orthografy?) to the VectorDataProvider by the commitChanges() method (include link to function) or by calling specific QgsVectorDataProvider functions"

It could be added to the following paragraph of https://qgis.org/pyqgis/master/core/QgsVectorLayer.html:

> The QgsVectorLayer is instantiated by specifying the name of a data provider, such as postgres or wfs, and url defining the specific data set to connect to. The vector layer constructor in turn instantiates a QgsVectorDataProvider subclass corresponding to the provider type, and passes it the url. The data provider connects to the data source. **Changes to the VectorLayer are syncronized to the VectorDataProvider by the commitChanges() method (include link to function)**

What do you think?

**Comment by tschmetzer** [comment]
> > feature.setAttributes([[0, 'examplename1'],[1,'10'],[2,'130']])
> 
> this is wrong -- you are setting your attributes to lists of two items each! I.e. the first attribute is being set to a list value of "[0, "examplename1"]

[...]

> ^^ This is the cause of your error. You're setting the attributes for debug_feature to a list of only 3 elements, whereas is should have 4 elements to match the size of the fields.
> 
> So then when you call:
> ```
> 
> debug_feature.setAttribute(debug_feature.fieldNameIndex(feld_type), layer_type)
> 
> ```
> You're calling:
> debug_feature.setAttribute(3, layer_type)
> Which is out of bounds for the 3 element list of attributes you just set, hence the error
> 
> What you should be doing is changing:
> 
> debug_layer.dataProvider().addFeature(debug_feature)
> to
> ```assert debug_layer.dataProvider().addFeature(debug_feature)```
> You'll see that the provider is rejecting the feature here, because it doesn't have the correct length of attributes set for the data provider. The message log window also contains a message describing why the addFeature called failed in detail.
> 
> So basically, you need to rework your field handling and insert more debugging statements into your code in order to track exactly what's occurring where!
> ```

Got it! Thank you so much for taking the time to explain! Apparently I was confused by the function specification of "setAttribute(self, field: int, attr: Any)" that seemed to use only indexes. Also I might not have been conscious about requiring to provide the *exact* number of field values. Yet when trying to use an assertion with a correct statement (e.g. `assert feature.setAttributes(['examplename1','10','130',None])` ) it failed. What do think of updating the docs like this, optionally update the function's code to facilitate the check for the exact number of required fields via an assertion.

https://qgis.org/pyqgis/master/core/QgsFeature.html#qgis.core.QgsFeature.setAttributes


>  setAttributes(self, attrs: object) → bool
> 
> 	Sets the feature’s attributes. The feature will be valid afterwards. **The number of provided attributes need to match exactly the number of the feature's fields.**
> 
> 	_(Optionally to facilitate an assertion)_
> 	_Returns False if number of attributes don't match exactly the number of the feature's fields._
> 
> 
> 	Parameters
> 		attrs (object) – List of **attribute values**
> 
> 	Returns
> 		_False if number of provided attribute values don't match exactly the number of the feature's fields_

**Comment by tschmetzer** [comment]
And I just realized that a 
`assert debug_layer.dataProvider().addFeature(debug_feature)`
doesn't seem to work:
- I can't see the method documented in https://qgis.org/pyqgis/master/core/QgsVectorDataProvider.html
- The method returns true in case of failure. (Yet a QgsMessageLog.logMessage is shown for attentive users:  `CRITICAL Layer debug_layer : Feature has too many attributes (expecting 32, received 33)`)

**Comment by nyalldawson** [comment]
@tschmetzer 
> The QgsVectorLayer is instantiated by specifying the name of a data provider, such as postgres or wfs, and url defining the specific data set to connect to. The vector layer constructor in turn instantiates a QgsVectorDataProvider subclass corresponding to the provider type, and passes it the url. The data provider connects to the data source. Changes to the VectorLayer are syncronized to the VectorDataProvider by the commitChanges() method (include link to function)

> What do you think?

Sounds good to me! How about "QgsVectorLayer implements a buffered approach to layer edits. Before edits can be made a call to startEditing() is required. Any edits made to a QgsVectorLayer are then held in memory only, and are not written to the underlying QgsVectorDataProvider until a call to commitChanges() is made. Buffered edits can be rolled back and discarded without altering the underlying provider by calling rollBack()."

> yet when trying to use an assertion with a correct statement (e.g. assert feature.setAttributes(['examplename1','10','130',None]) ) it failed.

QgsFeature.setAttributes() doesn't return anything, so you are effectively calling `assert None` here (which will always fail).
We could (should?) consider making setAttributes() raise a Python exception if the length of the attributes doesn't match with the feature's fields. That would certainly make debugging this issue much more straightforward.

> Sets the feature’s attributes. The feature will be valid afterwards. The number of provided attributes need to match exactly the number of the feature's fields.

That's a definite +1 ! You could format it nicely by going:

"\warning The number of provided attributes should exactly match the number of the feature's fields."

> I can't see the method documented in https://qgis.org/pyqgis/master/core/QgsVectorDataProvider.html

That's because QgsVectorDataProvider implements the QgsFeatureSink interface, which is where addFeature comes from

> In a script I'm writing the method returns true though it's a case of a failure. (Yet a QgsMessageLog.logMessage is shown for attentive users: CRITICAL Layer debug_layer : Feature has too many attributes (expecting 32, received 33). If I comment out the addFeature() statement then no error message is shown).

This sounds like a bug -- do you have a small script which demonstrates it?

**Comment by elpaso** [comment]
Please don't forget about transaction groups and passthrough buffer in the docs. That's an exception.
